substrate/docs/deprecation-process.md

```diff
@@ -0,0 +1,76 @@
+# Deprecation Process
+
+This deprecation process makes sense while we don’t use [SemVer](https://semver.org/).
+After that, this process will most likely change.
+As deprecation and removal of existing code can happen on any release, we need to be mindful that external builders
+could be impacted by the changes we make.
+The deprecation process tries to mitigate this impact, while still keeping the developer experience, the DevEx, as
+smooth as possible.
+
+To start a deprecation process a new issue with the label `I11-deprecation` needs to be created for correct tracking.
+Then these are the actions to take:
+
+## Hard deprecate by adding a warning message
+
+The warning message shall include a removal month and year, which is suggested to be 6 months after the deprecation
+notice is released.
```

Comment:
I would not recommend to set this time frame as a rule, it's good if your team is so organised that they know when they'll be ready to deprecate the code, but in some cases we can't really tell when this is gonna happen.. instead it's deprecated as soon as there's a replacement ready to merge

Reply:
> instead it's deprecated as soon as there's a replacement ready to merge

Yes, absolutely. The 6 months period is for the removal of the deprecated code, not for the deprecation itself.
When a piece of code is deprecated, it should be technically possible to be removed straight away. This period is for the downstream developers to have time to update their code. They should know when to expect the removal to happen.
Also, this timeframe (6 months or whatever was chosen) is only a lower bound, that's why the notice should look like `[DEPRECATED] will be removed after [DATE]`. This means that we are free to remove a code in 7 or 8 months even if we said that it will be removed after 6 months from now, but we shall not remove it before those 6 months.